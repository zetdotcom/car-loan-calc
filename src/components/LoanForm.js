import 'date-fns';
import React from 'react';
import Paper from '@material-ui/core/Paper';
import TextField from '@material-ui/core/TextField';
import Radio from '@material-ui/core/Radio';
import RadioGroup from '@material-ui/core/RadioGroup';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import FormLabel from '@material-ui/core/FormLabel';
import Button from 'components/ui/Button';
import SmallHeading from 'components/ui/SmallHeading';
import { MuiPickersUtilsProvider, DatePicker } from '@material-ui/pickers';
import DateFnsUtils from '@date-io/date-fns';

import styled from 'styled-components';
import { FormControl } from '@material-ui/core';

export const PaperStyled = styled(Paper)`
	display: flex;
	flex-direction: column;
	padding: 36px 48px;
	/* margin: 10px; */
	/* height: 300px; */
	/* min-height: 50vh; */
	justify-content: space-around;
	@media (max-width: 430px) {
		padding: 10px 15px;
		width: 100%;
	}
`;

const RadioGroupStyled = styled(RadioGroup)`
	&& {
		display: flex;
		flex-direction: row;
	}
`;

const FormStyled = styled.form`
	display: flex;
	flex: 1 1 300px;
`;

export default function LoanForm(props) {
	const {
		onSubmit,
		price,
		handlePriceChange,
		deposit,
		handleDepositChange,
		date,
		handleDateChange,
		loanTerm,
		handleLoanTermChange,
		depositError,
		dateError,
	} = props;

	console.log(date.getTime(), new Date().getTime(), 60 * 60 * 24);

	return (
		<FormStyled onSubmit={onSubmit}>
			<PaperStyled elevation={3}>
				<SmallHeading>Start loan calculation</SmallHeading>
				<TextField
					required
					type='number'
					margin='normal'
					id='price-input'
					label='Vehicle price'
					variant='outlined'
					value={price}
					onChange={handlePriceChange}
				/>
				<TextField
					required
					margin='normal'
					type='number'
					error={depositError}
					// helperText={depositError ? `Deposit value cannot be lower than 15% of price: ${price * 0.15}` : ''}
					helperText={`Deposit value must beat least 15% of price: ${price * 0.15}`}
					id='deposit-input'
					label='Deposit'
					variant='outlined'
					value={deposit}
					onChange={handleDepositChange}
				/>
				<MuiPickersUtilsProvider utils={DateFnsUtils}>
					<DatePicker
						error={dateError}
						inputVariant='outlined'
						helperText={dateError ? `Delivery date cannot be in the past` : ''}
						format='dd/MM/yyyy'
						margin='normal'
						id='date-picker'
						label='Delivery date'
						value={date}
						onChange={handleDateChange}
						InputAdornmentProps={{ position: 'start' }}
						KeyboardButtonProps={{
							'aria-label': 'change date',
						}}
					/>
				</MuiPickersUtilsProvider>
				<FormControl component='fieldset' margin='normal'>
					<FormLabel component='legend'>Loan Term</FormLabel>
					<RadioGroupStyled aria-label='Loan term' name='loan term' value={loanTerm} onChange={handleLoanTermChange}>
						<FormControlLabel value={1} control={<Radio />} label='1 year' />
						<FormControlLabel value={2} control={<Radio />} label='2 years' />
						<FormControlLabel value={3} control={<Radio />} label='3 years' />
					</RadioGroupStyled>
				</FormControl>
				<br />
				<Button variant='contained' color='primary' type='submit'>
					Show results
				</Button>
			</PaperStyled>
			{/* <div>
				price: {price}
				<br />
				deposit: {deposit}
				<br />
				date: {JSON.stringify(date)}
				<br />
				loan: {loanTerm} {typeof loanTerm}
			</div> */}
		</FormStyled>
	);
}
