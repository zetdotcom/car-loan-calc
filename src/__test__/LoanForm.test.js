import React from 'react';
import { render, cleanup, fireEvent, waitForElement } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import LoanForm from 'components/LoanForm';

describe('LoanForm', () => {
	const defaultProps = {
		onSubmit: jest.fn(),
		price: '',
		handlePriceChange: jest.fn(),
		deposit: '',
		handleDepositChange: jest.fn(),
		date: 'Thu Feb 20 2020 21:41:58 GMT+0000 (Greenwich Mean Time)',
		handleDateChange: jest.fn(),
		loanTerm: 1,
		handleLoanTermChange: jest.fn(),
		depositError: false,
	};

	const renderComponent = newProps => {
		const props = {
			...defaultProps,
			...newProps,
		};
		const utils = render(<LoanForm {...props} />);
		const component = <LoanForm {...props} />;
		return {
			component,
			...utils,
		};
	};

	afterEach(cleanup);
	it('should match snapshot', () => {
		const { asFragment } = renderComponent();
		expect(asFragment()).toMatchSnapshot();
	});
	it('should display correct form fields', async () => {
		const { getByLabelText, rerender } = renderComponent();
		const priceField = getByLabelText('Vehicle price *');
		// const priceField = getByTestId('priceInput');
		const depositField = getByLabelText('Deposit *');
		const dateField = getByLabelText('Delivery date');
		const checkbox1year = getByLabelText('1 year');
		const checkbox2years = getByLabelText('2 years');

		expect(priceField).toHaveTextContent('');
		expect(depositField).toHaveTextContent('');
		expect(dateField).toHaveValue('20/02/2020');
		expect(checkbox1year).toBeChecked();
		expect(checkbox2years).not.toBeChecked();

		rerender(renderComponent({ price: 1000, deposit: 200, loanTerm: 2 }).component);

		expect(priceField).toHaveValue(1000);
		expect(depositField).toHaveValue(200);
		expect(checkbox1year).not.toBeChecked();
		expect(checkbox2years).toBeChecked();
	});
});
