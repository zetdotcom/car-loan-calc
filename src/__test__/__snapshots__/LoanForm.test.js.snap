// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoanForm should match snapshot 1`] = `
<DocumentFragment>
  <form>
    <div
      class="MuiPaper-root sc-AxhUy gCIszr MuiPaper-elevation3 MuiPaper-rounded"
    >
      <h2
        class="sc-AxiKw hsrZvz"
      >
        Start loan calculation
      </h2>
      <div
        class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
          data-shrink="false"
          for="price-input"
          id="price-input-label"
        >
          Vehicle price
          <span
            class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
          >
             *
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input"
            id="price-input"
            required=""
            type="number"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-87 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-89"
            >
              <span>
                Vehicle price *
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
          data-shrink="false"
          for="deposit-input"
          id="deposit-input-label"
        >
          Deposit
          <span
            class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
          >
             *
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
        >
          <input
            aria-describedby="deposit-input-helper-text"
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input"
            id="deposit-input"
            required=""
            type="number"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-87 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-89"
            >
              <span>
                Deposit *
              </span>
            </legend>
          </fieldset>
        </div>
        <p
          class="MuiFormHelperText-root MuiFormHelperText-contained Mui-required"
          id="deposit-input-helper-text"
        >
          Deposit value must beat least 15% of price: £0.00
        </p>
      </div>
      <div
        class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal"
        format="dd/MM/yyyy"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined Mui-error Mui-error MuiFormLabel-filled"
          data-shrink="true"
          for="date-picker"
          id="date-picker-label"
        >
          Delivery date
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root Mui-error Mui-error MuiInputBase-formControl"
        >
          <input
            aria-describedby="date-picker-helper-text"
            aria-invalid="true"
            class="MuiInputBase-input MuiOutlinedInput-input"
            id="date-picker"
            readonly=""
            type="text"
            value="20/02/2020"
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-87 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-89 PrivateNotchedOutline-legendNotched-90"
            >
              <span>
                Delivery date
              </span>
            </legend>
          </fieldset>
        </div>
        <p
          class="MuiFormHelperText-root MuiFormHelperText-contained Mui-error MuiFormHelperText-filled"
          id="date-picker-helper-text"
        >
          Date should not be in the past
        </p>
      </div>
      <fieldset
        class="MuiFormControl-root MuiFormControl-marginNormal"
      >
        <legend
          class="MuiFormLabel-root"
        >
          Loan term
        </legend>
        <div
          aria-label="Loan term"
          class="MuiFormGroup-root sc-AxgMl jaInvp"
          role="radiogroup"
        >
          <label
            class="MuiFormControlLabel-root"
          >
            <span
              aria-disabled="false"
              class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-131 MuiRadio-root MuiRadio-colorSecondary PrivateSwitchBase-checked-132 Mui-checked MuiIconButton-colorSecondary"
            >
              <span
                class="MuiIconButton-label"
              >
                <input
                  checked=""
                  class="PrivateSwitchBase-input-134"
                  name="loan term"
                  type="radio"
                  value="1"
                />
                <div
                  class="PrivateRadioButtonIcon-root-147 PrivateRadioButtonIcon-checked-149"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    role="presentation"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                    />
                  </svg>
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root PrivateRadioButtonIcon-layer-148"
                    focusable="false"
                    role="presentation"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                    />
                  </svg>
                </div>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </span>
            <span
              class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
            >
              1 year
            </span>
          </label>
          <label
            class="MuiFormControlLabel-root"
          >
            <span
              aria-disabled="false"
              class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-131 MuiRadio-root MuiRadio-colorSecondary MuiIconButton-colorSecondary"
            >
              <span
                class="MuiIconButton-label"
              >
                <input
                  class="PrivateSwitchBase-input-134"
                  name="loan term"
                  type="radio"
                  value="2"
                />
                <div
                  class="PrivateRadioButtonIcon-root-147"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    role="presentation"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                    />
                  </svg>
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root PrivateRadioButtonIcon-layer-148"
                    focusable="false"
                    role="presentation"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                    />
                  </svg>
                </div>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </span>
            <span
              class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
            >
              2 years
            </span>
          </label>
          <label
            class="MuiFormControlLabel-root"
          >
            <span
              aria-disabled="false"
              class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-131 MuiRadio-root MuiRadio-colorSecondary MuiIconButton-colorSecondary"
            >
              <span
                class="MuiIconButton-label"
              >
                <input
                  class="PrivateSwitchBase-input-134"
                  name="loan term"
                  type="radio"
                  value="3"
                />
                <div
                  class="PrivateRadioButtonIcon-root-147"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    role="presentation"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                    />
                  </svg>
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root PrivateRadioButtonIcon-layer-148"
                    focusable="false"
                    role="presentation"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                    />
                  </svg>
                </div>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </span>
            <span
              class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
            >
              3 years
            </span>
          </label>
        </div>
      </fieldset>
      <br />
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained sc-AxjAm bixfS MuiButton-containedPrimary"
        tabindex="0"
        type="submit"
      >
        <span
          class="MuiButton-label"
        >
          Show results
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </form>
</DocumentFragment>
`;
